ios/engine/KMEI/KeymanEngine/Classes/Resource Management/ResourceDownloadQueue.swift

```diff
@@ -400,54 +436,54 @@ class ResourceDownloadQueue: HTTPDownloadDelegate {
-        if downloader!.requestsCount == 0 { // Download queue finished.
-          let keyboards = batch.resources as! [InstallableKeyboard]
-          log.info("Downloaded keyboard: \(keyboards[0].id).")
-
-          // TEMP:  wrap the newly-downloaded resources with a kmp.json.
-          //        Serves as a bridge until we're downloading actual .kmps for keyboards.
-          let wrappedKeyboards = Migrations.migrateToKMPFormat(keyboards)
-
-          wrappedKeyboards.forEach { keyboard in
-            ResourceFileManager.shared.addResource(keyboard)
-          }
-
-          if(!isUpdate) {
-            downloadSucceeded(forKeyboards: wrappedKeyboards)
-          }
-
-          let userDefaults = Storage.active.userDefaults
-          userDefaults.set([Date()], forKey: Key.synchronizeSWKeyboard)
-          userDefaults.synchronize()
+    let batch = queue.userInfo[Key.downloadBatch] as! AnyDownloadBatch
+    let isUpdate = batch.activity == .update
+
+    if let batch = batch as? DownloadBatch<InstallableKeyboard, KeyboardKeymanPackage> {
+      // The request has succeeded.
+      if downloader!.requestsCount == 0 { // Download queue finished.
+        let keyboards = batch.resources as! [InstallableKeyboard]
+        log.info("Downloaded keyboard: \(keyboards[0].id).")
+
+        // TEMP:  wrap the newly-downloaded resources with a kmp.json.
+        //        Serves as a bridge until we're downloading actual .kmps for keyboards.
+        let wrappedKeyboards = Migrations.migrateToKMPFormat(keyboards)
+
+        wrappedKeyboards.forEach { keyboard in
+          ResourceFileManager.shared.addResource(keyboard)
         }
-      } else if batch.type == .lexicalModel {
-        let task = batch.tasks[0] as! DownloadTask // It's always at this index.
-        if let lm = installLexicalModelPackage(downloadedPackageFile: URL(fileURLWithPath: task.request.destinationFile!)) {
-          if !isUpdate {
-            downloadSucceeded(forLexicalModel: lm)
-          } else {
-            // Since we don't generate the notification as above, we need to manually update
-            // the lexical model's metadata.
-            Manager.shared.updateUserLexicalModels(with: lm)
-          }
-        } else if !isUpdate {
-          let installError = NSError(domain: "Keyman", code: 0,
-                                     userInfo: [NSLocalizedDescriptionKey: "installError"])
-          downloadFailed(forLexicalModelPackage: "\(task.request.url)", error: installError)
-          currentFrame.batch?.errors[currentFrame.index] = installError
+
+        if(!isUpdate) {
+          downloadSucceeded(forKeyboards: wrappedKeyboards)
         }
+
+        let userDefaults = Storage.active.userDefaults
+        userDefaults.set([Date()], forKey: Key.synchronizeSWKeyboard)
+        userDefaults.synchronize()
+
+//        // TODO:  get package for keyboards.
+//        var package: KeyboardKeymanPackage
+//        batch.completionBlock?(package, nil)
```

Comment:
At the queue's core, we'll be able to return the downloaded package and all of its metadata.  Installation handling (and notification generation, after planned follow-up work) may then be taken from there.

Note that currently, all iOS resource installation assumes that the `InstalledLanguagesViewController` is loaded and present within the view hierarchy (because of the notification scheme, and because of how keyboard reloading currently works).  Part of the goal of this rework is to remove the former issue and allow the latter to be implemented as needed.

As a side note, this will get us closer to asynchronous / background update polling.